## Esercitazione 6  del    27 Marzo 2023 


## Esercizio 1 
Si consideri l'equazione quadratica $x^2 + 2px - q = 0 $ con $p=10^5$ e $q=10^{-i}$, i=1,..,10.
Si stabilisca se il problema relativo al calcolo della soluzione $x= -p+\sqrt{p^2+q}$ risulta essere ben
condizionato per tutti i valori di q assegnati;

In [1]:
import numpy as np
import sympy as sym
from sympy.utilities.lambdify import lambdify as lambdif

q = sym.symbols("q")
p = sym.symbols("p")

# array da 10**-1 a 10**-10
array = np.logspace(-1,-10,10)
print("Array: ",array)

f = (-p)+sym.sqrt(p**2+q)
df = sym.diff(f,q,1)
# print("Derivata prima di f: ",format(df))

fp=f.subs(p,10**5)
dfp=df.subs(p,10**5)
print(fp,dfp)

f_numerica = lambdif(q,fp, np)
df_numerica = lambdif(q,dfp, np)

k = np.abs(df_numerica(array)*array/f_numerica(array))
print("Array k: ",k)

Array:  [1.e-01 1.e-02 1.e-03 1.e-04 1.e-05 1.e-06 1.e-07 1.e-08 1.e-09 1.e-10]
sqrt(q + 10000000000) - 100000 1/(2*sqrt(q + 10000000000))
Array k:  [0.99999239 0.99999239 1.00174164 1.01058054 1.14532461 0.34359738
        inf        inf        inf        inf]


C:\Users\tommaso.ceredi\AppData\Local\Temp\ipykernel_13724\2192494800.py:23: RuntimeWarning: divide by zero encountered in divide
  k = np.abs(df_numerica(array)*array/f_numerica(array))


## Esercizio 2
Scrivere una funzione numpy che calcola la norma infinito e la norma 1 di un vettore ed una matrice  e testarla su vettori e matrici  a propria scelta. Confrontarne i risultati con quelli ottenuti utilizzando la funzione norm di numpy.linalg

(Ricorda la formula della norma infinito e della norma 1 di una matrice
$||A||_\infty= \max_{j=1,n} \sum_{i} |a_{ij}| $
$\quad ||A||_1= \max_{i=1,n} \sum_{j} |a_{ij}| $)

In [2]:
import numpy as np

def mynorm1(x):
    c = np.sum(np.abs(x), axis=0)
    n1 = np.max(c)
    return n1

def mynorm_inf(x):
    c = np.sum(np.abs(x), axis=1)
    n_inf = np.max(c)
    return n_inf

A = np.array([[1,2,3],[4,-5,6],[7,8,-19]])
norma1 = mynorm1(A)
norma1_py = np.linalg.norm(A, ord=1)
normainf = mynorm_inf(A)
normainf_py = np.linalg.norm(A, ord=np.inf)
print("norma1 mia: %d, norma1 numpy: %d" % (norma1, norma1_py))
print("normainf mia: %d, normainf numpy: %d" % (normainf, normainf_py))

norma1 mia: 28, norma1 numpy: 28
normainf mia: 34, normainf numpy: 34


## Esercizio 3
Implementare una funzione che calcola la norma 2 di una  matrice facendo uso della funzione eigvals del pacchetto numpy.linalg, (np.linalg.eigvals(A)). Testarla sulla matrice A=np.array([[4,-1,6],[2,3,-3],[1,-2,9/2]])   e confrontarne i risultati con quelli ottenuti utilizzando la funzione norm di numpy.linalg

In [3]:
import numpy as np

def mynorm2(A):
    c = np.sqrt(np.max(np.abs(np.linalg.eigvals(np.dot(A.T,A)))))
    return c

A=np.array([[4,-1,6],[2,3,-3],[1,-2,9/2]])

norma2 = mynorm2(A)
norma2_py = np.linalg.norm(A, ord=2)
print("norma2 mia: ",norma2, "norma2 numpy: ",norma2_py)

norma2 mia:  9.056251013341882 norma2 numpy:  9.056251013341878


## Esercizio 4 
Implementare il calcolo dell'indice di condizionamento in norma 2 di una matrice A come rapporto tra la radice dell'autovalore massimo di $A^TA$ e la radice dell'autovalore minimo di $A^TA$.

## Esercizio 5
- Costruire la matrice di Vandermonde A, generata dal vettore x=[1.0,2.0,...,6.0] utilizzando la funzione np.vander(x, increasing=True) di numpy. 
- Calcolarne l'indice di condizionamento in norma 1 senza utilizzare la funzione cond di numpy.linalg e poi confrontarne il valore con quello ottenuto utilizzando la funzione.
- Considerare il sistema lineare Ax=b  con matrice dei coefficienti A e termine noto costruito in maniera tale che la soluzione esatta sia il vettore x=[1,1,1,1,1,1] (ogni sua componente è 1)
- perturbare il vettore dei termini noti della quantità
$$
\delta b = 0.025 \, \ast \,
\left [
\begin{array}{c}
1\\
0\\
0\\
0
\end{array}
\right ]
$$
- Risolvere il sistema con termine noto pertubato $b + \delta b$ ((usando il metodo solve del modulo linalg di Scipy).
- Calcolare l'errore relativo sul termine noto e l'errore relativo sulla soluzione. Cosa si può concludere?

N.B. per il calcolo dell'inversa della matrice di A usare la funzione di numpy.linalg.inv(A).


In [4]:
import numpy as np

x = np.arange(1.0,7.0)
A = np.vander(x,increasing=True)
print(A)
mycond = mynorm1(A)*mynorm1(np.linalg.inv(A))
cond = np.linalg.cond(A, p=1)
print("Condizionamento: ",mycond, cond)

[[1.000e+00 1.000e+00 1.000e+00 1.000e+00 1.000e+00 1.000e+00]
 [1.000e+00 2.000e+00 4.000e+00 8.000e+00 1.600e+01 3.200e+01]
 [1.000e+00 3.000e+00 9.000e+00 2.700e+01 8.100e+01 2.430e+02]
 [1.000e+00 4.000e+00 1.600e+01 6.400e+01 2.560e+02 1.024e+03]
 [1.000e+00 5.000e+00 2.500e+01 1.250e+02 6.250e+02 3.125e+03]
 [1.000e+00 6.000e+00 3.600e+01 2.160e+02 1.296e+03 7.776e+03]]
Condizionamento:  1281104.9999999746 1281104.9999999746


## Esercizio 6
Assegnato il sistema lineare $Ax = b$, con
$$
A =
\left [
\begin{array}{ccc}
6 & 63 & 662.2\\
63 & 662.2 & 6967.8\\
662.2 & 6967.8 & 73393.5664
\end{array}
\right ], \qquad
b =
\left [
\begin{array}{c}
1.1\\
2.33\\
1.7
\end{array}
\right ]
$$
- trovare il vettore soluzione $x$ (usando il metodo solve del modulo linalg di Scipy);
- perturbare la matrice dei coefficienti della quantità
$$
\delta A =
0.01 \, \ast \,
\left [ \begin{array}{ccc}
1 & 0 & 0\\
0 & 0 & 0\\
0 & 0 & 0
\end{array}
\right ]
$$
quindi calcolare l'errore relativo sulla soluzione e confrontarlo con la
perturbazione relativa sui dati di ingresso. Cosa si osserva?

## Esercizio 7
 
Assegnato il sistema lineare $Ax = b$, con $A$ matrice di Hilbert di ordine
$4$ e $b = [1, 1, 1, 1]^T$,
 - trovare il vettore soluzione $x$ (usando il metodo solve del modulo linalg di Scipy);
 -  perturbare il vettore dei termini noti della quantità
$$
\delta b = 0.01 \, \ast \,
\left [
\begin{array}{c}
1\\
-1\\
1\\
-1
\end{array}
\right ]
$$
quindi calcolare la soluzione del sistema $A x_p= b_p$ con termine noto $b_p=b+ \delta b$.
Calcolare l'errore relativo sulla soluzione e confrontarlo con la perturbazione relativa sui dati di ingresso. Cosa si osserva?

Nota: per la costruzione della matrice di Hilbert usare la funzione hilbert(n) del modulo scipy.linalg
(scipy.linalg.hilbert(n))  dove bisogna specificare l'ordine n della matrice.

## Esercizio 8
Si consideri lo sviluppo in serie dell'esponenziale $e^x=1+x+\frac{1}{2!}x^2+\frac{1}{3!}x^3+\cdots =\sum_{k=0}^\infty \frac{x^k}{k!} $. Si calcoli un'approssimazione dell'esponenziale troncando ad N=100 termini la serie, per valori di x che variano da -10 a 10 incluso. Si calcoli  per ogni valore di x l'errore relativo del valore approssimato rispetto al valore esatto, ottenuto mediante la funzione exp della libreria numpy. Cosa si osserva? Pensare ad un algoritmo più stabile ricordando che $e^{-x}=\frac{1}{e^{|x|}}$. 

## Esercizio 9
Calcolare l'approssimazione della derivata prima di f(x) = sin(x) in x = 1
mediante l'algoritmo del rapporto incrementale (f(x + h) - f(x))/h per valori decrescenti
di h, confrontandolo con il valore fornito dalla funzione di libreria per f'(x)
mediante calcolo dell'errore relativo.